File: backend/app/schemas/panel_data.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


class PanelDataItem(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    province: str
    city: str
    year: int
    pv_installed_capacity_wan_kw: float | None = None
    disposable_income_per_capita_yuan: float | None = None
    healthcare_expenditure_per_capita_yuan: float | None = None
    urban_rural_income_ratio: float | None = None
    mortality_per_mille: float | None = None
    pm25_annual_avg_ug_per_m3: float | None = None
    gdp_100m_yuan: float | None = None
    source_sheet: str | None = None
    source_workbook: str | None = None


class PanelDataPage(BaseModel):
    items: list[PanelDataItem] = Field(default_factory=list)
    total: int = Field(default=0, ge=0)
    page: int = Field(default=1, ge=1)
    page_size: int = Field(default=20, ge=1)


class PanelDataPageResponse(BaseModel):
    status: Literal["success"] = "success"
    data: PanelDataPage


class PanelDataFilters(BaseModel):
    provinces: list[str] = Field(default_factory=list)
    cities: list[str] = Field(default_factory=list)
    years: list[int] = Field(default_factory=list)


class PanelDataFiltersResponse(BaseModel):
    status: Literal["success"] = "success"
    data: PanelDataFilters


class PanelDataProvinceStat(BaseModel):
    province: str
    count: int
    value: float


class PanelDataYearStat(BaseModel):
    year: int
    count: int
    value: float


class PanelDataStats(BaseModel):
    total_count: int = Field(default=0, ge=0)
    province_count: int = Field(default=0, ge=0)
    city_count: int = Field(default=0, ge=0)
    year_count: int = Field(default=0, ge=0)
    by_province: list[PanelDataProvinceStat] = Field(default_factory=list)
    by_year: list[PanelDataYearStat] = Field(default_factory=list)


class PanelDataStatsResponse(BaseModel):
    status: Literal["success"] = "success"
    data: PanelDataStats


class PanelDataMapItem(BaseModel):
    province: str
    city: str
    year: int
    value: float
    count: int
    longitude: float = Field(..., ge=-180, le=180)
    latitude: float = Field(..., ge=-90, le=90)


class PanelDataMapResponse(BaseModel):
    status: Literal["success"] = "success"
    data: list[PanelDataMapItem] = Field(default_factory=list)

File: frontend/src/api/panel_data.ts
import request from '@/api/request';

export interface PanelDataListParams {
  page?: number;
  page_size?: number;
  province?: string;
  city?: string;
  year?: number;
  keyword?: string;
}

export interface PanelDataFiltersParams {
  province?: string;
}

export interface PanelDataItem {
  id: number;
  province: string;
  city: string;
  year: number;
  pv_installed_capacity_wan_kw?: number | null;
  disposable_income_per_capita_yuan?: number | null;
  healthcare_expenditure_per_capita_yuan?: number | null;
  urban_rural_income_ratio?: number | null;
  mortality_per_mille?: number | null;
  pm25_annual_avg_ug_per_m3?: number | null;
  gdp_100m_yuan?: number | null;
  source_sheet?: string | null;
  source_workbook?: string | null;
}

export interface PanelDataPage {
  items: PanelDataItem[];
  total: number;
  page: number;
  page_size: number;
}

export interface PanelDataFilters {
  provinces: string[];
  cities: string[];
  years: number[];
}

export interface PanelDataStatsItem {
  province?: string;
  year?: number;
  count: number;
  value: number;
}

export interface PanelDataStats {
  total_count: number;
  province_count: number;
  city_count: number;
  year_count: number;
  by_province: Array<Required<Pick<PanelDataStatsItem, 'province'>> & Pick<PanelDataStatsItem, 'count' | 'value'>>;
  by_year: Array<Required<Pick<PanelDataStatsItem, 'year'>> & Pick<PanelDataStatsItem, 'count' | 'value'>>;
}

export interface PanelDataMapItem {
  province: string;
  city: string;
  year: number;
  value: number;
  count: number;
}

export function getPanelDataListApi(params?: PanelDataListParams): Promise<PanelDataPage> {
  return request({
    url: '/panel-data',
    method: 'get',
    params,
  }) as Promise<PanelDataPage>;
}

export function getPanelDataFiltersApi(params?: PanelDataFiltersParams): Promise<PanelDataFilters> {
  return request({
    url: '/panel-data/filters',
    method: 'get',
    params,
  }) as Promise<PanelDataFilters>;
}

export function getPanelDataStatsApi(): Promise<PanelDataStats> {
  return request({
    url: '/panel-data/stats',
    method: 'get',
  }) as Promise<PanelDataStats>;
}

export function getPanelDataMapApi(params?: Omit<PanelDataListParams, 'page' | 'page_size'>): Promise<PanelDataMapItem[]> {
  return request({
    url: '/panel-data/map',
    method: 'get',
    params,
  }) as Promise<PanelDataMapItem[]>;
}
